RFC: `--crate-attr`

jyn514 opened the pull request:
[Rendered](https://github.com/rust-lang/rfcs/blob/master/text/3791-crate-attr.md)

cc https://github.com/rust-lang/rust/issues/138287

Differences from the current nightly implementation (note that this section is non-normative):
- `-A -W -D -F` become aliases for `crate-attr`
- Attributes are prepended, not appended.
- Everything about how this interacts with doctests (see https://github.com/rust-lang/rfcs/pull/3791#discussion_r1999037718)
- The attribute shares the same Span as the crate root. In particular `file!` behaves properly now and the edition is no longer always 2015.

jyn514 committed:
RFC: `--crate-attr`

jyn514 commented on text/0000-crate-attr.md:
```diff
@@ -0,0 +1,113 @@
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+How should this interact with doctests? Does it apply to the crate being tested or to the generated test?
```
after thinking about this, i think this should always apply to the crate being tested. to apply to the generated test people can use `--crate-attr=doc(test(attr(...)))` (which is a bit of a mouthful, but, at least it's unambiguous.)

tamird replied:
such behavior would deviate from the semantics of existing flags like `--warn` which apply to *both* the crate being tested and its doctests.

jyn514 replied:
hmm, ok. i will change the RFC to specify it should apply to both.

cc @rust-lang/rustdoc

jyn514 replied:
i actually expanded this into something more complicated; see the RFC for details.

Nemo157 replied:
From what I can tell `rustdoc --test --deny=...` does not apply to either the crate being tested or its doctests. `rustdoc --test -Zcrate-attr` seems to only apply to the generated doctests crates, so `-Zcrate-attr=doc(test(attr(...)))` does not currently work.

jyn514 replied:
both of those seem like implementation bugs to me. i am happy to open an issue on rl/r if this RFC is merged.

epage commented on text/3791-crate-attr.md:
```diff
@@ -0,0 +1,113 @@
+3. Tools that require a specific attribute can pass that attribute automatically.
+
+Each of these corresponds to a different set of stakeholders. The first corresponds to developers writing Rust code. For this group, as the size of their code increases and they split it into multiple crates, it becomes more and more difficult to configure attributes for the whole workspace; they need to be duplicated into the root of each crate. Some attributes that could be useful to configure  ...
+- `no_std`
```
One of the potential routes for build-std is that the `Cargo.toml` has metadata for which std crates are used which means Cargo would need to be able to tell rustc about `#[no_std]`.  Unsure if there would have been another way to do it but this looks like it would allow that!

epage commented on text/3791-crate-attr.md:
```diff
@@ -0,0 +1,113 @@
+
+Additionally, some existing CLI options could have been useful as attributes. This leads to the second group: Maintainers of the Rust language. Often we need to decide between attributes and flags; either we duplicate features between the two (lints, `crate-name`, `crate-type`), or we make it harder to configure the options for stakeholder group 1.
+
+The third group is the authors of external tools. The [original motivation][impl] for this feature was for Crater, which wanted to enable a rustfix feature in *all* crates it built without having to modify the source code. Other motivations include the currently-unstable [`register-tool`], which with this RFC could be an attribute passed by the external tool (or configured in the workspace), and  ...
```
Looking at [`custom test frameworks`](https://github.com/rust-lang/rust/pull/52355), I'm not quite following how this would be useful.

Note that the current plan for custom test frameworks is very different than rust-lang/rfcs#2318, see rust-lang/testing-devex-team#2

jyn514 replied:
fair enough, i haven't been following test frameworks. i've removed this mention.

jyn514 committed:
initial feedback

- Mention how this applies to doctests
- Mention build-std
- Remove outdated mention of custom test frameworks

jyn514 committed:
forgot to mark doctests as resolved

jieyouxu commented on text/3791-crate-attr.md:
```diff
@@ -0,0 +1,121 @@
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `--crate-attr` flag allows you to inject attributes into the crate root.
+For example, `--crate-attr=crate_name="test"` acts as if `#![crate_name="test"]` were present before the first source line of the crate root.
```
Yeah that seems completely fair (if anything, may be QoI issue on specific crate-level attrs)

jieyouxu replied:
(I'll temporarily leave this thread open in case people do know of any Funny Things, but please resolve this comment chain before FCP as it is not intended to be blocking).

fmease replied:
I agree with Jyn that we shouldn't restrict the grammar of those attributes.

To answer the original question however, there are several. E.g., `#![doc = …]`, `#![deprecated = …]`, `#![type_length_limit = …]`. In these three cases you can use `concat!("", line!())` as the RHS which expands to (lineno) `"1"` when passed via `-Zcrate-attr` (I'm using concat to essentially get rid of the numeric suffix).

(`crate_name`, `crate_type` and `recursion_limit` are fully locked-down nowadays in the sense that they no longer accept macro calls)

(since a lot of malformed and misplaced attrs are still accepted by rustc and only trigger the warn-by-default `unused_attributes`, you can currently also do things like `#![must_use = compiler_error!("…")]` which isn't that interesting arguably)

Nemo157 replied:
`rustdoc -Zcrate-attr='doc=concat!("test", line!())' foo.rs` gave a `doc/foo/index.html`  containing "test1".

jyn514 replied:
ok. of these, `line!()` really does not seem to have a meaning to me, maybe the implementation comes up with "1" but i don't think that's useful. so i'm tempted to ban that particular macro in this context.

`include!` and friends all seem fine to me, they should be relative to whatever `#![doc = include_str!()]` in the source is relative to.

jyn514 replied:
i still think that all these concerns apply just as much to existing crate-level attributes in source code. if there is a concern here it should be specific to attributes passed *via a CLI flag*.

jyn514 commented on text/3791-crate-attr.md:
```diff
@@ -0,0 +1,121 @@
+
+The argument to `--crate-attr` is treated as-if it were surrounded by `#![ ]`, i.e. it must be an inner attribute and it cannot include multiple attributes, nor can it be any grammar production other than an [`Attr`].
+
+If the attribute is already present in the source code, it behaves exactly as it would if duplicated twice in the source.
+For example, duplicating `no_std` is idempotent; duplicating `crate_type` generates both types; and duplicating `crate_name` is idempotent if the names are the same and a hard error otherwise.
+It is suggested, but not required, that the implementation not warn on idempotent attributes, even if it would normally warn that duplicate attributes are unused.
```
so, it turns out i actually documented this wrong. the attributes are *appended* to the crate attrs in the source. what that means is crate-attr takes precedence over attributes in the source:
```
; head -n2 src/main.rs
#![warn(unused)]
fn foo() {}
; rustc +r2-stage1 src/main.rs -Zcrate-attr='allow(unused)' -Zcrate-attr='deny(unused)'

error: function `foo` is never used
 --> src/main.rs:4:4
  |
4 | fn foo() {}
  |    ^^^
  |
note: the lint level is defined here
 --> <crate attribute>:1:9
  |
1 | #![deny(unused)]
  |         ^^^^^^
```
this is unfortunate because it means that `crate-attr=allow` is not equivalent to `-A` (right now, the former takes precedence over the source, the latter does not).

i am tempted to say this is "just" a bug in an implementation and it should match the docs.

jieyouxu replied:
I can see why *this* order makes sense, but also I feel like `--crate-attr` taking precedence (or rather, get applied first) is "more useful" for actual users.

kpreid replied:
For the stated use case of “CLI flags are easier to configure for a whole workspace at once”, prepending is importantly more flexible because it allows crates to override the global configuration.

jieyouxu replied:
Ah I mean I can see why the *current* implementation order (appended) *might* be useful for a scenario where you *want* the cli crate-level attributes to come *after* the in-source crate-level attributes and thus take precedence *after* the in-source crate-level attributes. This is a very contrived example, but e.g.:

```bash
$ rustc --crate-attr='cfg_attr(target_os = "windows", allow(missing_docs))' foo.rs
```

```rs
#![warn(missing_docs)]
fn main() {}
```

With the *append*-ordering, you can use `--crate-attr` to suppress the lint but only if the `target_os` is windows.

(I'm personally in favor of the *prepend*-ordering, though it may be worth calling out this as an potential alternative behavior design choice.)

jyn514 replied:
it sounds like we've settled on the prepend ordering, but i want to call out explicitly: the point of this RFC is not for lints. there are already *many* ways to configure lints, including one explicitly supported by cargo. although i am documenting the behavior here, my hope is that no one should ever need to know or care about the order these are applied.

jyn514 committed:
Document that all tools support this flag. Document order and precedence.

jyn514 commented on text/3791-crate-attr.md:
```diff
@@ -0,0 +1,121 @@
+# Summary
+[summary]: #summary
+
+`--crate-attr` allows injecting crate-level attributes via the command line.
```
this flag is intended for all tools, including clippy and rustfmt, which i have now documented.

jyn514 commented on text/3791-crate-attr.md:
````diff
@@ -0,0 +1,134 @@
+fn foo() {}
+```
+
+This means that `crate-attr=deny(unused)` is exactly equivalent to `-D unused`.
````
oh, this is a good callout. i was imagining the semantics to be that -A/W/D flags are parsed before crate attrs are injected, which groups your examples the way that nightly's -Zcrate-attr currently does. but it would be cute to make A/W/D an alias for `crate-attr`, such that we can remove that flag processing in the compiler almost altogether and have `'warn(unused)' -A unused` mean `-A unused`. i'm in favor of that, i don't want to add yet more precedence grouping to lints.

jyn514 committed:
Define `-A/W/D` to be exactly equivalent to `--crate-attr`

jyn514 committed:
Clarify that `//!` isn't allowed

notriddle commented on text/3791-crate-attr.md:
```diff
@@ -0,0 +1,155 @@
+# Prior art
+[prior-art]: #prior-art
+
+- HTML allows `<meta http-equiv=...>` to emulate headers, which is very useful for using hosted infra where one does not control the server.
```
The [`set` built-in](https://www.gnu.org/software/bash/manual/html_node/The-Set-Builtin.html) for Bash fills a similar role, since every single-character option that can be configured with `set` can also be passed on the shell's command line.

jyn514 committed:
extend prior art section

jyn514 committed:
Document why crate-attr is prepended instead of appended

jyn514 committed:
Mention that `line!` and `column!` cause problems

jyn514 committed:
Mention that the more verbose syntax fixes column!

jyn514 committed:
oops forgot miri

jyn514 committed:
mention that #![crate_name] is a mess

jyn514 committed:
document that this shares a Span with the crate root

jyn514 committed:
Mention that the edition may be observable

jyn514 committed:
oops typo

jyn514 committed:
document that this may interact with custom inner attributes

jyn514 committed:
oops typo (redux)

jyn514 committed:
ok maybe edition should not actually be an attribute

Mark-Simulacrum commented on text/3791-crate-attr.md:
```diff
@@ -0,0 +1,181 @@
+
+1. The crate is parsed as if `--crate-attr` were not present.
+2. The attributes in `--crate-attr` are parsed.
+3. The attributes are injected at the top of the crate root.
```
```suggestion
3. The attributes are injected at the top of the crate root (see below for specifics on ordering).
```

Mark-Simulacrum commented:
@rfcbot fcp merge

I've wanted this functionality before myself, and it's a requested feature from Rust for Linux, so demand exists. I think the RFC does a good job motivating the feature and noting edge cases that need resolution, though personally nothing that necessarily seems like it blocks stabilization. So, proposing we merge it.

wesleywiser commented on text/3791-crate-attr.md:
```diff
@@ -0,0 +1,181 @@
+
+Additionally, some existing CLI options could have been useful as attributes. This leads to the second group: Maintainers of the Rust language. Often we need to decide between attributes and flags; either we duplicate features between the two (lints, `crate-name`, `crate-type`), or we make it harder to configure the options for stakeholder group 1.
+
+The third group is the authors of external tools. The [original motivation][impl] for this feature was for Crater, which wanted to enable a rustfix feature in *all* crates it built without having to modify the source code. Other motivations include the currently-unstable [`register-tool`], which with this RFC could be an attribute passed by the external tool (or configured in the workspace), and  ...
```
> and [build-std], which wants to inject `no_std` into all crates being compiled

Just as a note, RfL already does this exact thing and they would like to see `-Zcrate-attr` stabilized as well: 

https://github.com/Rust-for-Linux/linux/blob/e6ea10d5dbe082c54add289b44f08c9fcfe658af/scripts/Makefile.build#L238-L239

apiraino commented:
visited in meeting [on Zulip](https://rust-lang.zulipchat.com/#narrow/channel/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202025-04-03/near/509956361)

@rustbot label -I-compiler-nominated

Noratrieb commented on text/3791-crate-attr.md:
```diff
@@ -0,0 +1,181 @@
+
+- Is `--crate-name` equivalent to `--crate-attr=crate_name`? As currently implemented, the answer is no. Fixing this is hard; see https://github.com/rust-lang/rust/issues/91632 and https://github.com/rust-lang/rust/pull/108221#issuecomment-1435765434 (these do not directly answer why, but I am not aware of any documentation that does).
+
+- How should macros that give information about source code behave when used in this attribute? For example, `line!` does not seem to have an obvious behavior, and `column!` could either include or not include the surrounding `#![]`.
```
The problem presented here is the span of a `line!` *inside* a `--crate-attr`. So like a `--crate-attr doc(concat!("i'm on: ", line!()))`

nagisa replied:
I understand that. The load bearing part in my comment is the 2nd paragraph. Let me try rephrasing it.

If any non-trivial logic had to be prescribed to `line!()` and `column!()` inside these inserted attributes there are two ways to go about doing so:

1. Consider the inserted attributes as-if they were written in the source file. This then offsets every other `line!()` and/or `column!()` (thus breaking basic expectations of these macros); or
2. Drop the guarantee that a given `file:line:col` refers to exactly one location in the source.

Neither of these options are viable in my opinion. 1st, again, breaks basic expectations of these macros. As for 2nd option consider a situation where a diagnostic has to be shown at, say line 1 column 8~16. Does rustc display a snippet from an introduced attribute? A snippet from the source file before attributes were introduced? Maybe rustc isn't a great example, as error could be in *either* of these locations & rustc's source location tracking *will* need to be adjusted regardless. But what about other tools? Perhaps even tools that don't necessarily have any business parsing rust code (and thus no reason to accept `--crate-attr` as an argument,) but need to tie lines and columns back to the original source somehow?

These considerations ultimately lead to only one sensible implementation for `line!()`/`column!()` in introduced attributes: they both should result  in `0`, `1` or some similar placeholder of some sort that proper use of `line!()`/`column!()` cannot otherwise produce. Such approach can also be rationalized by describing attribute insertion as:

```
0 | insert_attributes!(); // attributes are being prepended to the source file & 0 comes before 1...
1 | // source code goes here, unperturbed...
```

Since macro invocations present a `#[track_location]`-like behaviour all `line!()` and `column!()` invocations that `insert_attributes!()` expands to produce the same value (in this case `0:0`, `1:1` or whatever else we pick.)

programmerjake replied:
why not just do what clang does and consider the inserted attributes to be in the file `<built-in>` or similar?
https://gcc.godbolt.org/z/aYn3cz5c3
`clang++ ... --include=cstddef --include=unknown2`
gives:
```
<built-in>:2:10: fatal error: 'unknown2' file not found
    2 | #include "unknown2"
      |          ^~~~~~~~~~
1 error generated.
Compiler returned: 1
```

Mark-Simulacrum replied:
I pushed an update removing the unresolved question and making an attempt at a decision on how this should work that seems in line with the suggestions here. Please take a look!

Mark-Simulacrum committed:
Clarify line!/column! treatment within expansion.

Noratrieb commented on text/3791-crate-attr.md:
````diff
@@ -0,0 +1,209 @@
+fn foo() {}
+```
+
+Additionally, all existing `-A -W -D -F` flags become aliases for `--crate-attr` (`allow`, `warn`, `deny`, and `forbid`, respectively). In particular, this implies that the following CLI flag combinations are exactly equivalent:
````
This doesn't actually affect anything, right? The flags behave exactly the same no matter whether they are aliases or not.

jyn514 replied:
no, this is user-facing. see https://github.com/rust-lang/rfcs/pull/3791#discussion_r1999189400 for an example of how it can affect behavior otherwise.

jyn514 commented on text/3791-crate-attr.md:
```diff
@@ -0,0 +1,209 @@
+[drawbacks]: #drawbacks
+
+It makes it harder for Rust developers to know whether it's idiomatic to use flags or attributes.
+In practice, this has not be a large drawback for `crate_name` and `crate_type`, although for lints perhaps a little more so since they were only recently stabilized in Cargo.
```
`more so` should be `moreso`. otherwise it is correct.

Noratrieb replied:
Oh, is "In practice, this has not be a large drawback" valid grammar? I don't think I've ever heard that expression, but I'm not a native speaker so idk^^

joshtriplett commented on text/3791-crate-attr.md:
```diff
@@ -0,0 +1,209 @@
+
+It makes it harder for Rust developers to know whether it's idiomatic to use flags or attributes.
+In practice, this has not be a large drawback for `crate_name` and `crate_type`, although for lints perhaps a little more so since they were only recently stabilized in Cargo.
+
```
```suggestion

Using this feature can make code in a crate dependent on attributes provided through the build system, such that the code doesn't build if reorganized or copied to another project.

```

(This drawback isn't meant to be a blocker, just an issue worth documenting.)

Mark-Simulacrum committed:
Document dependency on build system

Co-authored-by: Josh Triplett <josh@joshtriplett.org>
